Fix compilation on FreeBSD.

alexdupre opened the pull request:
This patch allows to successfully run `sandbox3/run` on FreeBSD.

This is not enough to have a fully working scala-native setup, but many applications will work.

There are at least a couple of additional issues that I've found:
- when the boehm GC is used with multithreading, FreeBSD requires linking with `gc-threaded` instead of just `gc`. I'm not sure how to change that. Moreover, the documentation should be updated unconditionally to require the `boehm-gc-threaded` package instead of `boehm-gc` (regardless the runtime use of multithreading)
- any jdk >= 11 on FreeBSD has the system param `java.net.preferIPv4Stack` set to `true` by default, and the IPv4-mapped IPv6 addresses are disabled by default. Many tests won't work unless the `net.inet6.ip6.v6only` sysctl is set to zero and `java.net.preferIPv4Stack` is set to `false`

ekrich commented:
> FreeBSD requires linking with gc-threaded instead of just gc

If this will work on all platforms, you could just change it here.
https://github.com/scala-native/scala-native/blob/main/tools/src/main/scala/scala/scalanative/build/GC.scala#L41

> any jdk >= 11 on FreeBSD

I am not sure here but it should be in the javadocs and should be true on all JVMs, not just FreeBSD for "write once run anywhere". If this is the case then I think an issue specific to this problem should be added. Others more familiar should weigh in.

alexdupre commented:
I don't think the change to link to`gc-threaded` will work on all platforms. My understanding is that FreeBSD provides a `boehm-gc` package without threads support, and a separate `boehm-gc-threaded` package with the multi-thread support that can both be installed concurrently, and for this reason the latter library has been renamed `gc-threaded`. I'm not sure what Linux distros do, I suppose they just provide a threaded library with the `gc` name. Can you confirm that?

I'll post another comment later about the networking issue. Anyway, these comments are just to let everyone be aware that there are still things to improve/fix, but they are independent from what this PR fixes.

LeeTibbert commented:
@alexdupre Thank you for the FreeBSD work here.  It would be nice to get FreeBSD
building and passing unit-tests.

TL; DR - There may be a fire here, but I believe it is not a barn fire.

The problems you report sound strange.

In January or so of 2023(?) I had FreeBSD building and passing almost all of the
unit-tests.  A few javalib network related tests did not pass (`NetworkInterface` etc.)
because the default configuration on FreeBSD differs from Linux & macOS.

Two or so months ago, I had a private FreeBSD building on, I think FreeBSD 13.0 (on AMD) (13.2?).
I bypassed the known failing network tests and made edits to a number of other files.
Unfortunately, I was never able to flush this work back to mainline (IIRC, a Scala version change
was two weeks away and I did not want to introduce churn.) 

I would have to review my work, but to my memory, no special linking was required.

I will have to re-check my now FreeBSD 14.0 system to see which JDK I was using, probably
Java 8.    

I will read thru the changes of this PR.

LeeTibbert commented on nativelib/src/main/resources/scala-native/delimcc.c:
```diff
@@ -15,11 +15,11 @@
 #define JMPBUF_STACK_POINTER_OFFSET (104 / 8)
 #elif defined(__x86_64__) &&                                                   \
-    (defined(__linux__) || defined(__APPLE__)) // x86-64 linux and macOS
+    (defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__)) // x86-64 linux, macOS and FreeBSD
```
I concur with adding `defined(__FreeBSD__)`.  

I checked my October 14 private porting and had made the same change.

Changing the comment is consistent with prior art (good!).
It is author's choice here. I deleted the entire comment rather that adding FreeBSD because the comment adds
no information beyond what the code says pretty explicitly
(due to good coding sytle).

LeeTibbert commented on nativelib/src/main/resources/scala-native/delimcc.c:
```diff
@@ -15,11 +15,11 @@
 #define JMPBUF_STACK_POINTER_OFFSET (16 / 8)
 #elif defined(__i386__) &&                                                     \
-    (defined(__linux__) || defined(__APPLE__)) // x86 linux and macOS
+    (defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__)) // x86 linux, macOS and FreeBSD
```
Has this change been tested on 32-bit FreeBSD?

By my understanding, 32 bit FreeBSD is fully supported
up to and including FreeBSD 12. In FreeBSD 13 and above, it became " Tier 2: Developmental and Niche Architectures".

If you are using & testing 32 bit FreeBSD, especially on FBSD > 12,  the change is fine.  If not, my suggestion is 
a separate branch which errors out on 32-bit FreeBSD (i.e. unsupported).   

That would allow testing and maintenance to focus on 64-bit, where such is struggling already.

ekrich replied:
Note that any changes that don't harm the mainline code towards the goal of first rate FreeBSD support should be welcome but I don't make that call. Also consider adding Setup documentation for FreeBSD and not supporting Boehm initially could be avoided by a simple "Boehm is not supported". Having a way via Docker or something to add to CI would be a great gain too even if some tests are "assumed" away. Even having one configuration that works would go a long way.

LeeTibbert replied:
Mostly agreed. Having an at least weekly FreeBSD CI run has been a goal (perhaps for a very small set of
people) since FreeBSD quasi-support was merged into mainline. I say "quasi" because it had no CI tests.

The problem has been that no one has had the time to make the required changes to CI.  

Given all the work that Wojciech has has put into rational CI runs over the past months/years,
the job should be easier, but such work still takes devo cycles.

LeeTibbert commented on nativelib/src/main/resources/scala-native/delimcc/setjmp_amd64.S:
```diff
@@ -1,5 +1,5 @@
 
-#if defined(__x86_64__) && (defined(__linux__) || defined(__APPLE__))
+#if defined(__x86_64__) && (defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__))
```
I concur with this change.

I never tried running on 32 bit FreeBSD.  If the intent is
to support and & test such (which I doubt SN should be doing in the absence of a champion/advocate). does the
corresponding setjmp_amd32.s need a corresponding change?

LeeTibbert commented on nativelib/src/main/resources/scala-native/gc/shared/Safepoint.c:
```diff
@@ -12,6 +12,9 @@
 #include <windows.h>
 #else // Unix
 #include <sys/mman.h>
+#ifndef MAP_NORESERVE
```
I suggest making this change FreeBSD specific.
That is, restrict the scope of variant handling to only
what we know & have exercised. 

Let future OSen add their own variant handling, if necessary.
< #if defined(__FreeBSD__)
< #if !defined(MAP_NORESERVE)
< #define MAP_NORESERVE 0 // concept does not exist on this FreeBSD; clear bit.
< #endif
< #endif

ekrich replied:
Perhaps a comment to indicate "FreeBSD" here might be good.

LeeTibbert reviewed (approved):
This PR Looks Good To Me (LGTM).

This PR contains the changes non-Test changes I made in private code in October, 2023 to get FreeBSD (13.2?) running on AMD hardware (well, VM on AMD hardware).

I expect the 14 or so Test files (unit-tests and `./tools/src/test/scala/scala/scalanative/linker/MinimalRequiredSymbolsTest.scala.original`) to continue to fail.
Having this PR merged will make feeding fixed FreeBSD 14 versions of those tests
back into mainline, perhaps a few at a time.



There are a couple of point suggestions, not requirements.  They should
not keep this PR from moving forward.



Because SN CI does not exercise FreeBSD (long story), it would be handy if the
base PR description described exactly __which__ FreeBSD versions and architectures
were tested.

LeeTibbert commented:
To co-ordinate work:

I expect at least the following tests to continue to fail.  I have fixed copies of many of these and will try
to peck away at creating PRs for them, perhaps a few at a time.
```
./tools/src/test/scala/scala/scalanative/linker/MinimalRequiredSymbolsTest.scala.original
./unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/LocaleTest.scala.original
./unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/MonetaryTest.scala.original
./unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/LocaleTest.scala.original~
./unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/ResourceTest.scala.original
./unit-tests/native/src/test/scala/org/scalanative/testsuite/posixlib/sys/StatTest.scala.original
./unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/InterfaceAddressTest.scala.original
./unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/NetworkInterfaceTest.scala.original
./unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/SocketTest.scala.original
./scripted-tests/run/build-library-dynamic/src/main/c/testlib.cpp.original
./scripted-tests/run/java-net-server-socket/tests/ServerTest.scala.original
./scripted-tests/run/java-net-socket/tests/EchoClientTest.scala.original
./scripted-tests/run/java-net-socket/test.original
```

LeeTibbert commented:
```
There are at least a couple of additional issues that I've found:

    when the boehm GC is used with multithreading, FreeBSD requires linking with gc-threaded instead of just gc. I'm not sure how to change that. Moreover, the documentation should be updated unconditionally to require the boehm-gc-threaded package instead of boehm-gc (regardless the runtime use of multithreading)

    any jdk >= 11 on FreeBSD has the system param java.net.preferIPv4Stack set to true by default, and the IPv4-mapped IPv6 addresses are disabled by default. Many tests won't work unless the net.inet6.ip6.v6only sysctl is set to zero and java.net.preferIPv4Stack is set to false

```
Could you create seperate Issues for each of these so that they do not get lost?  That will also make it easier
to get this PR merged (which will help in getting the tests, especially IPv6, working).

Each will require  focused study. The first issue deals with multithreading, which a whole other box.

Is the FreeBSD `boehm-gc-thread` package smart enough to fall back to (no-lock) straight line code when running
in a single-threaded process? Performance impacts?

alexdupre commented:
Some additional context:
1) I've been able to successfully run `sandbox3/run` on FreeBSD 13.2 amd64, FreeBSD 14.0 amd64 and FreeBSD 14.0 aarch64

2) I've tried to run `sandbox2_12/run` on FreeBSD 12.x i386, but it failed. I haven't a real i386 machine, so I tried within a jail. As you noted, to make it compile I had to add the `defined(__FreeBSD__)` into `setjmp_amd32.S` as well, but then it failed with:
```
[info] Generating intermediate code (with debug metadata) (1727 ms)
[info] Compiling to native code (565 ms)
[info] Linking native code (immix gc, none lto) (81 ms)
[info] Postprocessing (0 ms)
[info] Total (5583 ms)
Unexpected signal 11 when accessing memory at address 0x4000
        at _lh_get_sp
        at _lh_get_sp
        at _pthread_sigmask
        at _pthread_getspecific
[error] java.lang.RuntimeException: Nonzero exit code: 139
[error]         at scala.sys.package$.error(package.scala:30)
[error]         at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$18(ScalaNativePluginInternal.scala:276)
[error]         at scala.Option.foreach(Option.scala:407)
[error]         at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$16(ScalaNativePluginInternal.scala:276)
[error]         at scala.scalanative.sbtplugin.ScalaNativePluginInternal$.$anonfun$scalaNativeConfigSettings$16$adapted(ScalaNativePluginInternal.scala:253)
[error]         at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] stack trace is suppressed; run last sandbox2_12 / Compile / run for the full output
[error] (sandbox2_12 / Compile / run) Nonzero exit code: 139
```

so I guess FreeBSD/i386 is not actually supported. I don't see it as a problem, since it's now a tier2 platform and I don't think anyone will ever want to run scala-native on it, so we should probably create a separate branch which errors out in this scenario.

3) I couldn't run `tests3/test` on FreeBSD 14 aarch64 because of the following error, but I'm not sure why it happens:
```
[error] ld: error: undefined symbol: feraiseexcept
[error] >>> referenced by FenvTest.scala:59
[error] >>>               /home/ec2-user/scala-native/unit-tests/native/.3/target/scala-3.3.1/native-test/generated/6907d14c.ll.o:(_SM31scala.scalanative.libc.FEnvTestD46raiseCheckClearException$$anonfun$1$$anonfun$1L31scala.collection.immutable.ListuEpT31scala.scalanative.libc.FEnvTest)
[error] clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

4) on am64 I can start the unit tests, but some of them never terminate (CPU at 100% for hours on the same tests)

5) the sandbox program has been compiled with java 8 and java 21, the tests can be started only with java 8 because of the network differences between java 8 and 11+, specific to *BSD.

6) I've successfully compiled and run the sandbox program with all GC (none, boehm, immix, commix, experimental), but if I add `withMultithreadingSupport(true)` I get the following error:

```
[info] Compiling to native code (780 ms)
[error] ld: error: undefined symbol: __scalanative_loadModule
[error] >>> referenced by AbstractQueuedSynchronizer.scala:21
[error] >>>               /home/ale/scala-native/sandbox/.3/target/scala-3.3.1/native/generated/c957210c.ll.o:(_SM54java.util.concurrent.locks.AbstractQueuedSynchronizer$G4load)
[error] >>> referenced by RyuDouble.scala:38
[error] >>>               /home/ale/scala-native/sandbox/.3/target/scala-3.3.1/native/generated/7124a4ec.ll.o:(_SM56scala.scalanative.runtime.ieee754tostring.ryu.RyuDouble$G4load)
[error] >>> referenced by RyuFloat.scala:38
[error] >>>               /home/ale/scala-native/sandbox/.3/target/scala-3.3.1/native/generated/7124a4ec.ll.o:(_SM55scala.scalanative.runtime.ieee754tostring.ryu.RyuFloat$G4load)
[error] >>> referenced 87 more times
[error] ld: error: undefined symbol: scalanative_pthread_create
[error] >>> referenced by PosixThread.scala:72
[error] >>>               /home/ale/scala-native/sandbox/.3/target/scala-3.3.1/native/generated/ce284870.ll.o:(_SM26java.lang.impl.PosixThreadRL16java.lang.ThreadjE)
[error] clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

ekrich commented:
Given the differences in Java 8 vs Java 11, 17, and 21 I would document that FreeBSD should use Java 11+. Most projects are dropping Java 8 and there is no sense in trying to support old stuff any way. Scala Native still uses Java 8 in CI but I would be for dropping it entirely. 

I am sure you are doing `clean` between the builds of different configurations but thought I'd mention it just in case.

alexdupre commented:
I'm fine with requiring java 11+, but we need to find a good/agreed solution for the `TestMain`. The issue is that TestMain tries to connect to `::1` while the server is listening on tcp4 only. Simply removing the custom logic for FreeBSD, and trying to connect to 127.0.0.1 won't work either, because scala-native opens an AF_INET6 socket and BSD has disabled ipv6 mapped ipv4 addresses.

I see two solutions:
1) remove the custom logic and set the `java.net.preferIPv4Stack` system property instead (for FreeBSD only), so that scala-native will open a AF_INET socket to connect to 127.0.0.1
2) remove the custom logic and require the user to set `sysctl net.inet6.ip6.v6only=0`

I think the former is better, since it doesn't require any action by the user. It might break on an IPv6-only machine, though (not quite common).

LeeTibbert commented:
@alexdupre You have been busy & productive. Good to see.

There are a number of topic or work-items in this discussion so far.  I think we need some way of 
sorting them out and getting them inline (of priority).

My time is limited right now, but let me give some quick, and I hope helpful & encouraging, thoughts.

1) I concur that FreeBSD could drop 32 bit support. This, and other FreeBSD restrictions and options,
    should be described in the ReadTheDocs documentation for supported operating systems. I think
    the FreeBSD section was previously left vague on purpose because the edges of the FreeBSD world
    were not known.

2) re: restricting to JDK 11.  If it has to be on FreeBSD, then fine.  javalib does not have more 
    partially complete (50%, 60%) Java 8 support.  There are some Java > 8 parts implemented,
    but they are ~~almost useless~~ even further from completion.  Requiring Java 11 (or better yet 17) across the board
    would almost certainly lead to many well-deserved complaints about "missing methods".

3) I was waiting for this PR to merge before I started fixing unit-tests for FreeBSD.  
    I had them running in October. I believe that was with Java 8 and IPv6.  

    My suggestion is that we get this PR merged for Java 8, then correct the unit-tests on Java 8, 
    then run them on Java > 8 (probably 21, because most of my systems have that). Because
    there is currently no FreeBSD CI, this PR can be merged even if not everything is fixed at once.

     That would allow me to look at the `TestMain` code for FreeBSD in action on my systems
     and "one problem at a time".  

     It would also restore some working FreeBSD support whilst we work on other FreeBSD
     complexity.  

4)  If FreeBSD is setting `java.net.preferIPv4Stack=true` on JDK > mumble, that is more than passing
     strange.  I read/hear the report, but I need to understand the conditions (and possible FreeBSD discussions
     about the change.)  I'd like to understand this better before we make strategic decisions.

alexdupre commented:
I think to have addressed all your concerns/suggestions in my latest commits.

The change to TestMain should make it working OOTB on FreeBSD with stock config and any java version.

For the records, with these changes and the boehm-threaded hack I was able to successfully run all tests of https://github.com/lampepfl/gears

As noted in another github issue some native tests don't currently terminate on FreeBSD.


The reason to have "java.net.preferIPv4Stack=true" by default on BSD derives from the fact that IPv4-mapped IPv6 addresses are disabled by default for security reasons (https://datatracker.ietf.org/doc/html/draft-cmetz-v6ops-v4mapped-api-harmful-01). On OpenBSD there isn't even the ability to re-enable them, AFAIK.
The comment in this commit will give you some very useful insight: https://github.com/battleblow/jdk11u/commit/e1615c2ceac639ead79a7c3001720f468553ac36
Later it was extended to all BSD platforms and not just to OpenBSD: https://github.com/battleblow/jdk11u/commit/6e4f1d3de97900e545aca2000e2bbd7ec1cde43f
I'm not fully sure about the rationale behind such decision, but I think it's related to the fact that even when a dual mode socket was used, IPv4-mapped IPv6 addresses are still forbidden.

ekrich commented:
Please refer to our contributing docs for info about Java source code. https://scala-native.org/en/stable/contrib/contributing.html

LeeTibbert commented on docs/user/setup.rst:
```diff
@@ -93,16 +93,21 @@ installation of Arch Linux.
     $ sudo dnf install gc-devel zlib-devel # both optional
 
-**FreeBSD 12.2 and later**
+**FreeBSD 12.4 and later**
```
Edits of this section are well-said & appreciated.

LeeTibbert commented on docs/user/setup.rst:
```diff
@@ -93,16 +93,21 @@ installation of Arch Linux.
 .. code-block:: shell
 
-    $ pkg install llvm10
+    $ pkg install llvm # optional
```
Perhaps remove line entirely if acceptable version of 
llvm comes with FreeBSD 12.4?  Reduce cognitive load.

LeeTibbert commented on docs/user/setup.rst:
```diff
@@ -93,16 +93,21 @@ installation of Arch Linux.
     $ pkg install boehm-gc # optional
 
-*Note:* A version of zlib that is sufficiently recent comes with the
+*Note 1:* Sufficiently recent versions of llvm and zlib come with the
```
Perhaps move "boehm-gc" line after this paragraph,
then remove "Note" verbiage and promote to full paragraph? Then notes become renumbered.

alexdupre replied:
I'm not sure to fully understand how you'd like that I change it.

LeeTibbert replied:
I think this simplifies down to something like:

```
**FreeBSD 12.4 and later**
*Note 1:* Only AMD64 and ARM64 architectures are supported.
*Note 2:* Sufficiently recent versions of llvm and zlib come with the
installation of FreeBSD.

.. code-block:: shell

    $ pkg install boehm-gc # optional

*Note 3:* Using the boehm GC with multi-threaded binaries doesn't work
out-of-the-box yet.
```

Puts the important information (only 64 bit) first and
keeps optional  boehm-gc info close together in layout.

What do you think?  Feel free to reject or wordsmith.

alexdupre commented:
> Please refer to our contributing docs for info about Java source code.

If you are referring to the Scala CLA I think I already signed it in the past. Anyway, I just signed it again.

ekrich commented:
I was mostly referring to the links supplied which seem to be pointing to OpenJDK code.

alexdupre commented:
> I was mostly referring to the links supplied which seem to be pointing to OpenJDK code.

Well, those links have just the purpose to show the jdk code running on BSD platforms, no code has been brought into this PR.

LeeTibbert commented on test-interface/src/main/scala/scala/scalanative/testinterface/TestMain.scala:
```diff
@@ -31,53 +31,27 @@ object TestMain {
   final val iPv6Loopback = "::1"
 
-  private def getFreeBSDLoopbackAddr(): String = {
+  private def setFreeBSDWorkaround(): Unit = {
```
The "not allowing IPv4 mapped IPv6 addresses," does not match my memory of what was working for me circa 2023-10-15.   When I take a run, hopefully soon, at fixing
unit-tests for FreeBSD, I will have to refresh my memory.

I believe I was running most unit-tests with an unmodified
TestMain on IPv6.  If that was and continues to be true, this section may need to be re-worked.

alexdupre replied:
The fact that you had to introduce `iPv6Loopback` and use it on FreeBSD (even if openjdk8 on FreeBSD *has* dual mode sockets) was exactly because IPv4 mapped IPv6 addresses are forbidden, so using `127.0.0.1` on an `AF_INET6` socket wasn't working.
If you enable them, then on Java8  there isn't anything else to do, but on java 11+ the server will continue to listen only on tcp4, unless `java.net.preferIPv4Stack` is explicitely set to false on sbt. Manually adjusting both settings seems much more complex than automatically  push SN to use an IPv4 socket.

LeeTibbert commented on test-interface/src/main/scala/scala/scalanative/testinterface/TestMain.scala:
```diff
@@ -31,53 +31,27 @@ object TestMain {
-     * Tests which use IPv4 addresses, either through hard-coding or
-     * bind resolution, on IPv6 systems will still fail. This allows to
-     * run the vast majority of tests which do not have this characteristic.
+     *   sysctl net.inet6.ip6.v6only=0
```
s/by running/by previously running/  

or such: i.e. no concurrent modification

alexdupre replied:
Hmmm, what do you mean with concurrent modification?

LeeTibbert replied:
I am suggesting making the timing of the "sysctl" explicit.  That is, before TestMain is run.
That should give the desired results. i.e. sysctl & TestMain do not run at the same time, probably
in different processes.

As you may know better than I, sometimes changing a sysctl value will take effect immediately while a 
program is executing and sometimes only the next time it is run.  I suspect that changing ip6.v6only
"on-the-fly" will either have no effect or mess TestMain up totally.  

Yes, "on-the-fly is allowed" is a perverse interpretation of the existing text. Since the existing text
does not rule it out, somebody is bound to try it. Then again, even with it being ruled out, somebody
is bound to try it ;-)

LeeTibbert commented on test-interface/src/main/scala/scala/scalanative/testinterface/TestMain.scala:
```diff
@@ -31,53 +31,27 @@ object TestMain {
+     *
+     * The easiest way to make TestMain to work on most FreeBSD machines
+     * with different Java versions without changing system defaults is
+     * to set java.net.preferIPv4Stack=true in Scala Native in order to
```
s/to set/ to set, __before__ the first Java network call,/

or such. In SN, the check for preferIPv4Stack is done once,
before the first Java network call.

LeeTibbert commented on test-interface/src/main/scala/scala/scalanative/testinterface/TestMain.scala:
```diff
@@ -31,53 +31,27 @@ object TestMain {
   final val iPv4Loopback = "127.0.0.1"
   final val iPv6Loopback = "::1"
```
This val seems to be unused, probably for a long time.

Although it goes against my usual "don't change unless the change is __needed_ for __this__ PR" approach,
I think deleting the line (if I am correct about it being unused), helps a maintainer trace if and how IPv6 
might be being used.  

On known IPv6 systems, using the IPv6Loopback is likely to be slightly more efficient. I had thought that
once upon a time, this file did so.

alexdupre replied:
See comment above, I'll remove both variables and restore the direct use of `127.0.0.1`

LeeTibbert commented on scripts/scalafmt:
```diff
@@ -3,7 +3,7 @@
 
 HERE="`dirname $0`"
-VERSION=$(sed -nre "s#\s*version[^0-9]+([0-9.]+)#\1#p" $HERE/../.scalafmt.conf)
+VERSION=$(sed -nre "s#version[^0-9]+([0-9.]+)#\1#p" $HERE/../.scalafmt.conf)
```
I'm curious & for posterity trying to figure this change out: Why does deleting matching optional 
whitespace make things work on FreeBSD (and is
not required on Linux, etc.).

There is a group of people who like optional whitespace and another which vehemently oppose it.   In the days of punched cards, whitespace was expensive.

alexdupre replied:
Using `\s` to skip whitespaces doesn't seem portable. To keep the behavior I can replace it with `[[:space:]]`, i.e. to `"s#[[:space:]]*version[^0-9]+([0-9.]+)#\1#p"`

alexdupre replied:
```
% sed -nre "s#\s*version[^0-9]+([0-9.]+)#\1#p" .scalafmt.conf
sed: 1: "s#\s*version[^0-9]+([0- ...": RE error: trailing backslash (\)

% sed -nre "s#[[:space:]]*version[^0-9]+([0-9.]+)#\1#p" .scalafmt.conf
3.7.14
```

LeeTibbert replied:
I like the "[[:space:]]" approach.   

Do you have a Windows system?  If yes, does it work there?  If no, I can dust of my Windows system, find 
its mouse & keyboard, fire it up, and try it there.

alexdupre replied:
It works on gnu sed, and so it seems to work on windows, too.

LeeTibbert commented on docs/user/setup.rst:
```diff
@@ -93,16 +93,21 @@ installation of Arch Linux.
+
+*Note 3:* Using the boehm GC with multi-threaded binaries doesn't work
+out-of-the-box yet.
+
```
Given that a number of Tests, primarily unit-tests, are known to not execute, it may be worthwhile to add a "Note 4" which says something like:
"
A number of Tests, primarily unit-tests, are known to fail
on FreeBSD.  It is believed that the code-under-test is correct and that the defect is in the Test itself.

Work is underway to fix these tests. See [PR](https://github.com/scala-native/scala-native/pull/3625) for a list of tests known to fail at that time."

I/we could create an Issue, with checkboxes, which lists
the known failing tests  (Java 8) and substitute that URL into the text above. 

That way, we are saying what we are doing and doing what
we say.

The last PR in the series which fixes these tests would delete this Note.  I usually dislike creating  work items which have to be deleted later, but this work is in flux.

Your thoughts?

alexdupre commented:
I've pushed some other improvements, I think I'm done.

LeeTibbert commented:
@alexdupre 

Thank you for the info about FreeBSD dual stack.

LeeTibbert commented:
I think this PR is almost done.  I'd like to give Wojciech, who will probably be the one doing the merge,
a clear "Yes, this is ready" in a final reply here.  That would mean that Wojciech would only have to
lightly read a productive but long conversation trail.  

To move it PR forward and capture unresolved concerns, I suggest creating at least two formal
Issues (in addition to the "hanging unit-tests" entry").  That will allow focused attention on each.

1) One which says something like:
```
mumble
FreeBSD 12.4+
java > 11
IPv4-only, dual-stack?
mumble.  
```

2) multi-thread boehm-gc link failure on BSD.

alexdupre commented:
Why do you think the IPv4/dual-stack issue is still unresolved? What would be the purpose of the new github issue?

I've contacted the boehm-gc maintainer today and I hope he'll change the package to install the threaded libraries with the standard name, as done by other OS, so no change will be needed by SN. I'm waiting for his feedback.

LeeTibbert commented:
> Why do you think the IPv4/dual-stack issue is still unresolved? What would be the purpose of the new github issue?

The idea is to convince the person doing the merge is that "due diligence" has been done.

If there is an OS specific hack necessary, I would like to understand it. Not the least because such will
probably come back to haunt me.    

I have read the reports of dual-stack dragon encounters and believe them. Forewarned is forearmed.

Having an Issue, from a person other than me, would help keep the issue from falling off the bottom of my 
"I'd like to fix that" priority list.  I am suggesting a 5-minutes-work place holder, probably referring to this PR,
not a three-day effort.    I  can create it if you like. 

> I've contacted the boehm-gc maintainer today and I hope he'll change the package to install the threaded libraries with the standard name, as done by other OS, so no change will be needed by SN. I'm waiting for his feedback.

Thank you for doing that.  Please keep posted.  It would be nice to not have to do anything here.

LeeTibbert commented on scripts/scalafmt:
```diff
@@ -3,7 +3,7 @@
 
 HERE="`dirname $0`"
-VERSION=$(sed -nre "s#version[^0-9]+([0-9.]+)#\1#p" $HERE/../.scalafmt.conf)
+VERSION=$(sed -nre "s#[[:space:]]*version[^0-9]+([0-9.]+)#\1#p" $HERE/../.scalafmt.conf)
```
Sorry to repeat myself.  Has this been run manually on Windows?
We, you & I, are going to win no friends if we break SN development on
Windows.

CI runs on Ubuntu Linux. Unless we know that this has been successfully manually run on Windows, we should assume it broken.

As mentioned, I can run this on Windows if such would be hard for you.

alexdupre replied:
I'm not sure what's the build env on windows. I've tried on mingw64 and it worked, I don't know what exact `sed` implementation uses SN on Windows. To be 100% sure it's better if you try yourself.

LeeTibbert replied:
Now that I look at this more closely, I realize that I am a goat, the barnyard critter, not the 
"Greatest of all time" kind.

This is obviously a bash script, which will not run on plain Windows to begin with. Arrgh!  :facepalm:        
It might be run on WSL (linux on Windows), but any `sed` there should handle the POSIX
`[[:space:]]`

Sorry for the false trigger.

alexdupre commented:
https://github.com/scala-native/scala-native/issues/3629
https://github.com/scala-native/scala-native/issues/3630

LeeTibbert commented:
CI failure summary

TL;DR  I believe the two failures and the several multi-arch warnings on Windows are unrelated to
            the content of this PR.

The following breakage in this also occurs in merged PR #3626. Breakage may have entered
the main branch.

3 Multiarch warnings 
`"Exception calling "SetPageFileSize" with "3" argument(s): "The operation completed successfully" `

2 runtime failures, TestMain loopback connection failing, but only on Scala 2 & Windows. 
```
Exception in thread "main"java.net.SocketException: read failed, errno: 10054
	at java.net.AbstractPlainSocketImpl.read(Unknown Source)
	at java.net.SocketInputStream.read(Unknown Source)
Exception in thread "main"java.net.SocketException: read failed, errno: 10054
	at java.net.AbstractPlainSocketImpl.read(Unknown Source)
```

LeeTibbert commented:
@WojciechMazur 

I believe this PR is ready for review & merge at your convenience.

There is a long & complicted conversation trail for this PR.  The issues raised and not resolved in 
the current PR are FreeBSD or BSD (possibly macOS) specific and should not affect Linux & Windows. 
A formal SN Issue has been created for each, so they are not lost.

A number of `unit-tests` are known to fail on FreeBSD. That has an issue and is described
as a limitation in the `setup.rst` documentation for FreeBSD.  From prior private work,
I believe these are failures in the Tests themselves, and not the underlying code.
Such things as OS specific device names and such.

When this PR is merged, I will peck away at merging my private unit-test fixes.

WojciechMazur reviewed (approved):
Looks good to me. It's great to see restored balance in the FreeBSD support. Hopefully one day we're be able to set up a CI to guard it continuously.
